benchmarks/persistence/persistence.py

```diff
@@ -0,0 +1,235 @@
+    """The boto3 session object. This will load the respective credentials
+    from the environment variables within the container."""
+
+    @_bucket_name.default
+    def _default_bucket_name(self) -> str:
+        """Get the bucket name from the environment variables."""
+        if not PERSIST_DATA_TO_S3_BUCKET:
+            raise ValueError(
+                "No S3 bucket name provided. Please provide the "
+                + "bucket name by setting the environment variable "
+                + "BAYBE_BENCHMARKING_PERSISTENCE_PATH."
+            )
+        return os.environ["BAYBE_BENCHMARKING_PERSISTENCE_PATH"]
+
+    @override
+    def write_json(self, object: dict, path_constructor: PathConstructor) -> None:
+        """Store a JSON serializable dict according to the path.
+
+        This method will store an JSON serializable dict in an S3 bucket.
+        The S3-key of the Java Script Notation Object will be created by
+        the path object. I the key already exists, it will be overwritten.
+
+        Args:
+            object: The object to be persisted.
+            path_constructor: The PathConstructor to create the path for the object.
+
+        """
+        client = self._object_session.client("s3")
+
+        key = path_constructor.get_path(strategy=PathStrategy.HIERARCHICAL)
+
+        client.put_object(
+            Bucket=self._bucket_name,
+            Key=key.as_posix(),
+            Body=json.dumps(object),
+            ContentType="application/json",
+        )
+
+
+@define
+class LocalFileSystemObjectWriter(ObjectWriter):
+    """Class for persisting JSON serializable dicts locally."""
+
+    folder_path_prefix: Path = field(validator=instance_of(Path), default=Path("."))
+    """The prefix of the folder path where the results are stored.
+    The filename will be created automatically and create or override the
+    file under this path. The file path must exist."""
+
+    @override
+    def write_json(self, object: dict, path_constructor: PathConstructor) -> None:
+        """Store a JSON serializable dict according to the path.
+
+        This method will store an JSON serializable dict in a local file system.
+        If the respective file exists, it will be overwritten.
+
+        Args:
+            object: The object to be persisted.
+            path_constructor: The PathConstructor to create the path for the object.
+
+
+        Raises:
+            FileNotFoundError: If the folder path prefix does not exist.
+        """
+        if not self.folder_path_prefix.exists():
+            raise FileNotFoundError(
+                f"The folder path {self.folder_path_prefix.resolve()} does not exist."
+            )
+        path_object = self.folder_path_prefix.joinpath(
+            path_constructor.get_path(strategy=PathStrategy.FLAT)
+        )
+        with open(path_object.resolve(), "w") as file:
+            json.dump(object, file)
+
+
+def make_object_writer() -> ObjectWriter:
```

Comment:
I dislike the `make_` approach and we also often follow different patterns that are equally valid choices here (the `make_` approach does not seem very needed here too imo)

Not sure why `make_object_writer` is even needed, given that we dont need local storage persistence and Im not sure why it was even implemented (see other comment)

`make_path_constructor` would become obsolete if `PathConstructor` would just hold a `benchmark` and `result` as attributes and the current constructor attribtutes were properties

Reply:
That separation comes from a former code version. I think it would be fine to drop it. The `PathConstructor` class may also have a constructor which performs the work of  `make_path_constructor`. The `make_object_writer` was my way to hide the choice of which class is used for writing, that can also be changed but I think local writing is good for testing purposes. @AdrianSosic What do you think?

Reply:
1. Regarding `make_path_constructor`: removing it is fine for me 👍🏼 As you correctly pointed out, you can take the code exactly as-is and turn it into a corresponding `from_...` `classmethod` with appropriate name in the `PathConstructor` class as alternative constructor. So basically just attaching the logic to the class instead of standalone function. While we *could theoretically* also change the class such that it directly takes the method's inputs as attributes, I vote against it since the `attrs` mantra `you should think about what the class needs and not how it is going to be instantiated` has proven incredibly useful in the past 🙃 
2. I also consider the possibility of being able to run stuff locally *very* useful. So **somewhere** this distinction needs to be made. But you can also just replace the function with a simple ternary `if-else` operator in the `__main__.py`, which actually might be more elegant given the current simplistic logic

Reply:
Okay, changed it.

Reply:
Looking good to me 👍🏼 But will let @Scienfitz resolve since he opened the thread

Reply:
local running will probably be practically useless unless you want to let your machine run days on the side, and theres also a different between running things locally and additionally persisting them. But given after the last changes theres no silent CI fail anymore im fine with it